### PR TITLE
Add Hunter support to 4.0.4

### PR DESCRIPTION
**Pull request by caseymcc**

### CODE REVIEW

File: cmake/SetupAndInstallPdb.cmake [hunk 1]
```diff
@@ -0,0 +1,72 @@
+cmake_minimum_required(VERSION 3.2.3) #using the continue instruction
+
+macro(SetupAndInstallPdb)
```
Comment (ruslo) [inline cmake/SetupAndInstallPdb.cmake:3]: I don't know much details but just for your info I'm using this approach:
* https://github.com/forexample/package-example/blob/00335ff3f39c3b82a4b4b6095412b5a9e3aeda81/Foo/CMakeLists.txt#L24-L85

File: CMakeLists.txt [hunk 2]
```diff
@@ -14,17 +15,28 @@ IF(COMMAND CMAKE_POLICY)
   endif(POLICY CMP0022)
 ENDIF(COMMAND CMAKE_POLICY)
 
-PROJECT(WT)
+option(WT_USE_HUNTER "Enable Hunter CMake package handler, auto downloads and compiles require packages" OFF)
```
Comment (ruslo) [inline CMakeLists.txt:18]: Why not `HUNTER_ENABLED`?

See:
* https://docs.hunter.sh/en/latest/reference/user-variables.html#hunter-enabled
* https://docs.hunter.sh/en/latest/overview/compatibility.html

File: CMakeLists.txt [hunk 3]
```diff
@@ -144,14 +149,26 @@ IF(WIN32)
...
 
-  IF(SHARED_LIBS)
+  IF(BUILD_SHARED_LIBS)
     # See http://svn.boost.org/trac/boost/ticket/3465
     SET(WT_NO_BOOST_INTRUSIVE true)
-  ENDIF(SHARED_LIBS)
+  ENDIF(BUILD_SHARED_LIBS)
+
+  #Set build directory
+  set( CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin/ )
```
Comment (ruslo) [inline CMakeLists.txt:158]: What is the reason to touch these directories? Usually you don't have to touch them.

File: cmake/WtFindBoost-hunter.txt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+FOREACH(_target ${Boost_COMPONENTS})
+  IF(TARGET Boost::${_target})
+    STRING(TOUPPER ${_target} _TARGET)
+    SET(Boost_${_TARGET}_LIBRARY Boost::${_target})
+  ENDIF(TARGET Boost::${_target})
+ENDFOREACH(_target ${Boost_COMPONENTS})
+UNSET(_target)
+UNSET(_TARGET)
+
+IF (Boost_FOUND)
```
Comment (ruslo) [inline cmake/WtFindBoost-hunter.txt:43]: This variable will be always `TRUE`, you can do:
```cmake
if(NOT Boost_FOUND)
  message(FATAL_ERROR "...")
endif()
```

File: cmake/WtFindPng.txt [hunk 5]
```diff
@@ -6,41 +6,47 @@
...
-      ${PNG_PREFIX}/lib
-      ${USERLIB_PREFIX}/lib
-  )
-ENDIF(WIN32)
+if(WT_USE_HUNTER)
+    hunter_add_package(PNG)
+    find_package(PNG REQUIRED)
+    set(PNG_LIBRARIES PNG::png)
+else(WT_USE_HUNTER)
+    IF(NOT PNG_NAMES)
```
Comment (ruslo) [inline cmake/WtFindPng.txt:14]: You can violate formatting rules for sake of minimal change:
```cmake
if(WT_USE_HUNTER)
  # injected code
else()

# old code, without indentation, untouched

endif()
```

in this particular case I think you can do:
```cmake
if(WT_USE_HUNTER)
  # injected code
  return() # <<< exit from module
endif()

# old code, untouched
```

**Review by ruslo — COMMENTED** [review]
If you want to patch upstream 4.0.4 version you should create branch `hunter-4.0.4` starting from this tag, see:
* https://docs.hunter.sh/en/latest/creating-new/patch.html

I've created branch `hunter-4.0.4`:
* https://github.com/hunter-packages/wt/tree/hunter-4.0.4

It's better to send patches there.

> https://github.com/hunter-packages/wt/releases/tag/4.0.4

You should name tags like `4.0.4-pN` or `4.0.4-hunter-N` so it will be clean that it's not an upstream tag `4.0.4`.

File: cmake/SetupAndInstallPdb.cmake (hunk 1 above)
Comment (caseymcc) [reply cmake/SetupAndInstallPdb.cmake:3]: yeah, i used your example for reference. I just wanted something I could use universally. Is there something about the script that you think could be better?

Comment (ruslo) [reply cmake/SetupAndInstallPdb.cmake:3]: > Is there something about the script that you think could be better?

I just noticed that using approach from https://github.com/forexample/package-example you don't have to add the code for every target, you just need to set few variables in top level CMakeLists.txt.

Comment (caseymcc) [reply cmake/SetupAndInstallPdb.cmake:3]: but this only installs the pdb for the targets that were manually added and it follows the CMAKE_\<CONFIG\>_POSTFIX.

File: CMakeLists.txt (hunk 2 above)
Comment (caseymcc) [reply CMakeLists.txt:18]: I was trying to get this back into the mainline so I was trying to make sure HunterGate did not have to be included, but I doubt that is going to happen so likely its better to just use it normally.

File: cmake/WtFindPng.txt (hunk 5 above)
Comment (caseymcc) [reply cmake/WtFindPng.txt]: Ill modify it and then merge with the 4.0.4-p0 branch I created (and revert this merge).

**Comment by caseymcc** [comment]
Ill modify it and then merge with the 4.0.4-p0 branch I created (and revert this merge).

File: CMakeLists.txt (hunk 2 above)
Comment (ruslo) [reply CMakeLists.txt:18]: > I was trying to get this back into the mainline so I was trying to make sure HunterGate did not have to be included

HunterGate will do nothing if `HUNTER_ENABLED` is `OFF`, that's what "Backward compatibility" section is about.

**Comment by ruslo** [comment]
> Ill modify it and then merge with the 4.0.4-p0 branch I created (and revert this merge).

Branch should be called `hunter-4.0.4` (already created) and tag should be called `4.0.4-p0`.

**Pull request merged**